### PR TITLE
[Polymer UI] Show success notification on file operations

### PR DESCRIPTION
**Pull request by yebrahim**
This change uses the recently added toast notification to show a success message when a file operation succeeds. A big part of this change is white spaces for the file upload change so I recommend using the `?w=1` querystring.

image

### CODE REVIEW

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts [hunk 1]
```diff
@@ -392,90 +392,87 @@ class FilesElement extends Polymer.Element {
...
-          okLabel: 'Upload Anyway',
-        };
-
-        const result: BaseDialogCloseResult = 
-          await Utils.showDialog(BaseDialogElement, dialogOptions);
-          proceedWithUpload = result.confirmed === true;
-      }
+      const result: BaseDialogCloseResult = 
+        await Utils.showDialog(BaseDialogElement, dialogOptions);
+        proceedWithUpload = result.confirmed === true;
```
Comment (blois) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: Indentation?

Any progress on a lint/fix?

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts [hunk 2]
```diff
@@ -392,90 +392,87 @@ class FilesElement extends Polymer.Element {
...
-            reader.readAsDataURL(file);
-          });
-
-          // Now upload the file data to the backend server.
-          const uploadPromise = loadPromise
-            .then((model: JupyterFile) => ApiManager.saveJupyterFile(model));
-          uploadPromises.push(uploadPromise);
-        }
-      }
+    if (proceedWithUpload) {
```
Comment (blois) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: Why not return early? Would make the code more readable.

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts [hunk 3]
```diff
@@ -392,90 +392,87 @@ class FilesElement extends Polymer.Element {
...
-          title: 'Warning: Large File',
-          messageHtml: warningMsg,
-          okLabel: 'Upload Anyway',
-        };
-
-        const result: BaseDialogCloseResult = 
-          await Utils.showDialog(BaseDialogElement, dialogOptions);
-          proceedWithUpload = result.confirmed === true;
-      }
+      const result: BaseDialogCloseResult =
```
Comment (blois) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: Random: is this type needed?

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts [hunk 4]
```diff
@@ -392,90 +392,87 @@ class FilesElement extends Polymer.Element {
...
+              format: 'base64',
+              content: itemData,
+            };
+
+            resolve(model);
+          }
+
+          // TODO: handle file reading errors.
+          reader.onerror = () => {
+            reject('Error reading file.');
```
Comment (blois) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: Should always reject with Errors, as these are directly translated to throws.

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts [hunk 5]
```diff
@@ -392,90 +392,87 @@ class FilesElement extends Polymer.Element {
...
-          const uploadPromise = loadPromise
-            .then((model: JupyterFile) => ApiManager.saveJupyterFile(model));
-          uploadPromises.push(uploadPromise);
-        }
-      }
+    if (proceedWithUpload) {
+      files.forEach((file: File) => {
+
+        // First, load the file data into memory.
+        const loadPromise = new Promise((resolve, reject) => {
```
Comment (blois) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: Nit: promise creation should be as narrowly scoped as possible to make it easier to validate that it is always resolved.

Would recommend just:

```
const loadPromise = new Promise((resolve, reject) => {
  const reader = new FileReader();
  reader.onload = () => reader.result;
  reader.onerror = (e) => throw new Error('Error reading file');
  reader.readAsDataURL(file);
});
```

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts [hunk 6]
```diff
@@ -392,90 +392,87 @@ class FilesElement extends Polymer.Element {
...
+
+          const reader = new FileReader();
+
+          reader.onload = function() {
```
Comment (blois) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: This should probably be an arrow function, as it is elsewhere in this method.

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts [hunk 7]
```diff
@@ -392,90 +392,87 @@ class FilesElement extends Polymer.Element {
...
+        });
+
+        // Now upload the file data to the backend server.
+        const uploadPromise = loadPromise
+          .then((model: JupyterFile) => ApiManager.saveJupyterFile(model));
+        uploadPromises.push(uploadPromise);
+      });
 
       // Wait on all upload requests before declaring success or failure.
       return Promise.all(uploadPromises)
```
Comment (blois) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: It's an async function- why not use that?

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts [hunk 8]
```diff
@@ -392,90 +392,87 @@ class FilesElement extends Polymer.Element {
    */
   async _upload() {
```
Comment (blois) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: Why is all of this logic in the UI layer? Really seems like it should be at a lower level.

Tests?

Comment (yebrahim) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: The polymer element contains all of the UI as well as underlying logic. Plus, this logic here is all UI related, it's mostly building the warning message if any and displaying it for large files.

I'm afraid tests will have to come as a separate change.

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts (hunk 3 above)
Comment (yebrahim) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: I'd like to keep it for clarity. Even thought `showDialog` returns a `BaseDialogCloseResult`, this is a base type, and the actual type returned will differ according to the dialog type passed. So it's just a coincidence that we really are expecting a `BaseDialogCloseResult` in this case, and I think it's worth being explicit about that.

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts (hunk 1 above)
Comment (yebrahim) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: Fixed and sent in a separate PR.

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts (hunk 2 above)
Comment (yebrahim) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: Done.

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts (hunk 5 above)
Comment (yebrahim) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: Done.

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts (hunk 6 above)
Comment (yebrahim) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: Done.

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts (hunk 4 above)
Comment (yebrahim) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: Done.

File: sources/web/datalab/polymer/components/datalab-files/datalab-files.ts (hunk 7 above)
Comment (yebrahim) [inline sources/web/datalab/polymer/components/datalab-files/datalab-files.ts]: Done.
